lib/Transforms/InsertGpuAllocs.cpp

```diff
@@ -0,0 +1,363 @@
+        -> llvm::Optional<mlir::SmallVector<mlir::Value, 4>> {
+      if (auto load = mlir::dyn_cast<mlir::memref::LoadOp>(op)) {
+        return {{load.memref()}};
+      } else if (auto store = mlir::dyn_cast<mlir::memref::StoreOp>(op)) {
+        return {{store.memref()}};
+      } else if (auto call = mlir::dyn_cast<mlir::func::CallOp>(op)) {
+        mlir::SmallVector<mlir::Value, 4> ret;
+        for (auto arg : call.operands()) {
+          if (arg.getType().isa<mlir::MemRefType>())
+            ret.emplace_back(arg);
+        }
+        return std::move(ret);
+      } else {
+        op->emitError("Uhhandled mem op in gpu region");
+        return llvm::None;
+      }
+    };
+
+    auto scfDialect = getContext().getOrLoadDialect<mlir::scf::SCFDialect>();
+
+    auto hasMemAccess = [](mlir::Operation *op) -> bool {
+      if (auto memInterface =
+              mlir::dyn_cast<mlir::MemoryEffectOpInterface>(op)) {
+        if (memInterface.hasEffect<mlir::MemoryEffects::Read>() ||
+            memInterface.hasEffect<mlir::MemoryEffects::Write>())
+          return true;
+      }
+      if (auto call = mlir::dyn_cast<mlir::func::CallOp>(op)) {
+        for (auto arg : call.operands()) {
+          if (arg.getType().isa<mlir::MemRefType>())
+            return true;
+        }
+      }
+      return false;
+    };
+
+    // Traverse through all the memory access ops under GPU launch Op
+    // and add device memory allocation appropriately.
+    if (func.walk([&](mlir::Operation *op) {
+              if (!op->getParentOfType<mlir::gpu::LaunchOp>())
+                return mlir::WalkResult::advance();
+
+              if (!hasMemAccess(op))
+                return mlir::WalkResult::advance();
+
+              auto memref = getMemref(op);
+              if (!memref)
+                return mlir::WalkResult::interrupt();
+
+              for (auto mem : *memref) {
+                while (auto parentView =
+                           mem.getDefiningOp<mlir::ViewLikeOpInterface>())
+                  mem = parentView.getViewSource();
+
+                for (auto alias : aliases.resolve(mem)) {
+                  auto op = alias.getDefiningOp();
+                  if (op) {
+                    if (mlir::isa<mlir::memref::GetGlobalOp>(op)) {
+                      gpuGetMemrefGlobalParams.insert({op, {}});
+                      continue;
+                    }
+                    if (op->getDialect() == scfDialect ||
+                        mlir::isa<mlir::ViewLikeOpInterface>(op))
+                      continue;
+                    if (mlir::isa<mlir::memref::AllocOp>(op)) {
+                      gpuBufferAllocs.insert({op, {}});
+                    } else if (mlir::isa<mlir::func::CallOp>(op)) {
+                      // Ignore
+                    } else {
+                      op->emitError("Unhandled memref producer");
+                      return mlir::WalkResult::interrupt();
+                    }
+
+                  } else {
```

Comment:
what is the condition for this? a "non" op?

Reply:
this is the gpu params case,

Reply:
>>this is the gpu params case,
Here you may want to put an assert statement to assert your assumption. 

The logic is not straightforward - 

Here you assume that if alias.getDefiningOp() do return a null pointer,  then the assumption is that alias can give Parentblocks where you can find the arguments.  What if the loop body has nested loop inside? Do you write test case to verify?